### PR TITLE
Fresh commit of "Create open registration E2E example""

### PR DESCRIPTION
**Pull request by bdhirsh**
cc @albanD

Main changes from the test in core are:
(1) Slightly beefed up comments
(2) Updated kernels just to print when they're called (and added the output of running the script in the comments)
(3) Showed two example user API's for a custom device: added a TorchFunctionMode that lets you pass in `device="foo"` and have things work properly

### CODE REVIEW

File: open_registration_example.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+print("Creating x on device 'foo:0'")
+x1 = torch.ones(4, 4, device='foo:0')
+print("Creating y on device 'foo:0'")
+y1 = torch.ones(4, 4, device='foo:0')
+
+test(x1, y1)
+
+# Normally you would probably want this device TorchFunction translation to happen
+# permanently, bu this example I want to remove it.
```
Comment (albanD) [inline open_registration_example.py]: but

File: cpp_extensions/open_registration_extension.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  m.impl("_copy_from", &custom__copy_from);
+}
+
+// This basic implementation doesn't bother dealing with different device indices
+// (e.g. custom_device:0 vs. custom_device:1).
+// We could do that by letting the user pass in a device index in our exposed device function.
+// Note that if you do that, you'll also need to register a device guard to core.
+// See `c10/core/impl/DeviceGuardImplInterface.h:C10_REGISTER_GUARD_IMPL`.
+c10::Device get_custom_device(int idx) {
+  return c10::Device(c10::DeviceType::PrivateUse1, idx);
```
Comment (albanD) [inline cpp_extensions/open_registration_extension.cpp:271]: How do you actually specify which indices are valid?

**Review by albanD — COMMENTED** [review]
Sounds good!
Maybe also add a top level README?

File: cpp_extensions/open_registration_extension.cpp (hunk 2 above)
Comment (bdhirsh) [reply cpp_extensions/open_registration_extension.cpp:271]: good point - I ended up adding an example `DeviceGuard` in this PR (pushing changes in a minute), that has the barebones logic to error out if you pass an invalid index.

**Pull request merged**